Fix infiniteScroll in tableView

juzen2003 opened the pull request:
- This is related to #762 
  - When scrolling slider in table view, the slider has the step size equal to gallery row size.
  - When scrolling ps in table view, the slider will be updated base on gallery row size.
  - When switching between browse tab and other tabs, the slider will stay at the same value.
  - When we move the slider to the end (max value), scroll down the ps to the loadNext threshold point (near the bottom of ps), and switching between gallery and table views, the slider value will stay the same.

juzen2003 committed:
- In browse.js (line 1505), modify the condition to update the obsNum in
  infiniteScroll instances only when obsNum is smaller than the result
  count.
- In browse.js (line 732), modify updateSliderHandle function to remove
  the condition check "if (o_browse.isGalleryView())" so that slider will
  move by the gallery step size in table view.

juzen2003 committed:
- In opus.js, create getCurrentTab function to return the current tab that
  the user selects (line 503-508).
- In browse.js, ,modify updateSliderHandle to only update the slider and
  obsNum in infiniteScroll instances when the current tab is browse tab
  (line 677-681).

rfrenchseti commented on opus/application/static_media/js/browse.js:
```diff
@@ -674,6 +674,11 @@ var o_browse = {
 
     // find the first displayed observation index & id in the upper left corner
     updateSliderHandle: function(browserResized=false) {
+        // Only update the slider & obSnum in infiniteScroll instances when the user
```
What is the point of this new function? Doesn't it do exactly the same thing as getViewTab below?

juzen2003 replied:
This is directly reading from opus.prefs.view and it will have possible returned value equal to "search", "browse", "cart", and "detail". It will also let us know where users are at right now, for example, it will return "search" if the user is at search page. getViewTab only returns "#cart" or "#browse", and if the user is at search page, the function still returns "#browse".  The reason of using this function is to make sure updateSliderHandle only gets called when we select browse tab (or we are at browse tab). This is to avoid updateSliderHandle being called from adjustBrowseHeight when mutationObserver detects a tab changes. For example, when we are in browse view and click search nav-link, mutationObserver will detect a change and call adjustBrowseHeight even though we are getting out of browse view. By adding the check here, updateSliderHandle will not be called when we are exiting out the browse tab, so the startObs stored in the infiniteScroll instances and startobs in url will not get messed up (because the calculation in updateSliderHandle is based on the existing elements in browse tab, variables like topBoxBoundary will get messed up in above example).

rfrenchseti replied:
Ah, I see. I think we might want to rename "getViewTab" to "getViewBrowseCartTab" or something to avoid confusion.

rfrenchseti reviewed (changes requested):
The scrollbar still doesn't work well in the table view. Sometimes when you scroll down, the new page loads, and the scrollbar jumps back to the middle, the slider also jumps (as well as the top row of the table jumping back). You can probably reproduce this by sliding to a place so you clear the cache, then slowly pressing the down arrow to move one row at a time. When you get to the place where the scrollbar resets to the middle, the slider will move back substantially as well.

I've also seen cases where the top of the table is several gallery rows off from what the slider says.

juzen2003 committed:
- In browse.js (line 1508-1514), properly update the obsNum in infiniteScroll
  instances so that table view scrollbar and slider are in the correct location
  after infiniteScroll load event is triggered.

juzen2003 committed:
- In browse.js (line 612-615), update setScrollbarPosition to take in two
  obsNum parameters, one for gallery view and another one for table view,
  they are used to set scrollbar location for gallery/table view.
- In browse.js, update infiniteScroll instances to store "sliderObsNum" and
  "scrollbarObsNum". "sliderObsNum" will be the startObs. "scrollbarObsNum"
  will be the obsNum at current scrollbar location. In table view, it's
  the top item, in gallery view, it's one of the item in the top row.

juzen2003 committed:
- In browse.js (line 752-754), add the definitions of "sliderObsNum" and
  "scrollbarObsNum" as comments.
